### ISSUE TITLE
Use dataset snapshot to generate ingest import

### ISSUE THREAD

**Issue by amnonkhen** [ebi-ait/dcp-ingest-central#329]
Using [Dataset Tracking Sheet - Project Catalogue "Todo" Snapshot; Jun 3rd 2021](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=1934812232), populate ingest with new projects, and review.

#### Import Group 1
- [x] Ask Devs if we can use the DevDB for reviewing import results.
- [x] Ask Wranglers if we can use Wrangling Priority / Wrangling Status to flag projects that need review.
- [x] Get Json representation of snapshot
- [x] except where duplicates would be created (by uuid or doi)
- [x] build project metadata from tracking sheet columns
- [x] Add data from DOI lookup
- [x] post new projects to ingest (on dev for testing)
- [x] save ingest uuid's back to tracker (manual intervention is fine)
- [x] Import Organs and Technology ~~(Using: [this script](https://github.com/ebi-ait/hca-ebi-dev-team/tree/master/scripts/set-project-tech-organs))~~
- [x] @ray @clairerye to populate DOI's from pub_link
- [x] Resolve [Organ Ontologies that cannot be found](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=1923571015)
- [x] Resolve [Technology Ontologies that cannot be found](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=746177999)
- [ ] Resolve [Ontologies that do not match exactly](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=781850240)
- [x] Test import of Group 3 @MightyAx
- [ ] Resolve (or ignore) Group 3 Ontology Issues @Wkt8
    - [2 missing ontology](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit?disco=AAAAMvWECZg)
    - [7 non exact matches](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit?disco=AAAAMvWECZQ)
- [ ] Raise "exceptions" with @rays22

[Original Tracking Sheet](https://docs.google.com/spreadsheets/d/1rm5NZQjE-9rZ2YmK_HwjW-LgvFTTLs7Q6MzHbhPftRE/edit#gid=0)

**Comment by MightyAx** [comment]
Don't include "unsuitable"

**Comment by MightyAx** [comment]
Useful Fields:
Identifying uniqueness (new project or updating existing project)

- ingest_project_uuid
- doi

Filling project details:

- access_permission
- organism
- sample_type
- assay_type
- organ
- nucleic_acid_source

**Comment by MightyAx** [comment]
Also useful: this google sheet:
https://docs.google.com/spreadsheets/d/1bE9fbMw0Ujgjag3odL1bnbtqKCw_43T_AHz5mebrzyc/edit#gid=0

which used google script to query ingest and return the results in the sheets cells:
```
function setLink() {
  var spreadsheet = SpreadsheetApp.getActive();
  var cur_cell = spreadsheet.getCurrentCell();
  var link = cur_cell.getRichTextValue().getLinkUrl();
  cur_cell.setRichTextValue(SpreadsheetApp.newRichTextValue()
  .setText(link)
  .build());
  spreadsheet.getCurrentCell().offset(1, 0).activate();
  setlink();
};

function setProjectValidation() {
  var spreadsheet = SpreadsheetApp.getActive();
  var link = spreadsheet.getCurrentCell().getRichTextValue().getLinkUrl();
  var http_response = UrlFetchApp.fetch(link);
  var json = JSON.parse(http_response);
  spreadsheet.getCurrentCell().offset(0, 6).activate();
  spreadsheet.getCurrentCell().setRichTextValue(SpreadsheetApp.newRichTextValue().setText(json.validationState).build());
  spreadsheet.getCurrentCell().offset(1, -6).activate();
  setProjectValidation();
};

function setSubmissionValidation() {
  var spreadsheet = SpreadsheetApp.getActive();
  var link = spreadsheet.getCurrentCell().getRichTextValue().getLinkUrl();
  link = link + "/submissionEnvelopes"
  var http_response = UrlFetchApp.fetch(link);
  var json = JSON.parse(http_response);
  var result = json.page.totalElements + " Submissions"; 
  spreadsheet.getCurrentCell().offset(0, 7).activate();
  spreadsheet.getCurrentCell().setRichTextValue(SpreadsheetApp.newRichTextValue().setText(result).build());
  spreadsheet.getCurrentCell().offset(1, -7).activate();
  setSubmissionValidation();
};
```

**Comment by tburdett** [comment]
I've taken a snapshot of the data tracking sheet and grouped it according to what next steps should be, see https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=0

This gives a total cohort of projects for the Catalogue at 432 (which may need to change slightly if we find discrepancies)

@MightyAx I have not changed the structure of the sheets at all, column structures etc are identical so if you were working on a programmatic solution to import it should work equally well on the new doc

**Comment by tburdett** [comment]
As discussed, @MightyAx is going to work through the projects in group 1 of the above sheet, @rays22 will review outcomes and decide on next steps for any discrepancies

**Comment by MightyAx** [comment]
Ways to distinguish imported projects:

- Wrangling Priority: 3
- Notes: "Imported from Snapshot (Date)"

**Comment by gabsie** [comment]
@rays22 @ESapenaVentura @Wkt8 @ami-day  - Hi, wranglers, for all the imported projects we will place them as Priority 3 for now, and will include in the notes the above message "Imported from Snapshot"

Let us know if that is OK for you, 
Gabs

**Comment by rays22** [comment]
> @rays22 @ESapenaVentura @Wkt8 @ami-day - Hi, wranglers, for all the imported projects we will place them as Priority 3 for now, and will include in the notes the above message "Imported from Snapshot"
> 
> Let us know if that is OK for you,
> Gabs

Yes, the default values for priority and notes look sensible to me, so it is is fine with me.

**Comment by ESapenaVentura** [comment]
Yep, that's fine with me as well :)

**Comment by rays22** [comment]
### Review of `2021-06-03 Snapshot.json` of **Group 1**

* 10 of the 19 ignored projects in the `2021-06-03 Snapshot.json` are already are in Ingest and they have project UUIDs in the [Dataset Tracking Sheet snapshot.](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=0)

```
ingest_project_uuid	dcp_id
027c51c6-0719-469f-a7f5-640fe57cbece	dcp7082152978536813
fc55bd4a-8694-4a28-9a35-59a685bda323	dcp6376670064684554
08fb10df-32e5-456c-9882-e33fcd49077a	dcp1669513903415692
2bda737d-bd81-412f-b37b-cf76b059f778	dcp9396450730750780
2d4d89f2-ebeb-467c-ae60-a3efc5e8d4ba	dcp9881798822220041
379ed69e-be05-48bc-af5e-a7fc589709bf	dcp0867377478600200
6936da41-3692-46bb-bca1-cd0f507991e9	dcp3363310633713612
b43c36dd-c8ec-4cf1-a8f4-6ab5f30eabc3	dcp9250113319871051
e579d0f4-18ef-4d42-90fa-4d2afb60a862	dcp2254658349927023
fc232d87-7b10-47f7-9197-b2c6cf210dea	dcp6883522079544929
```

* I should note that two of the ten Ingest project pages are crashing my browser:
    1. [027c51c6-0719-469f-a7f5-640fe57cbece](https://contribute.data.humancellatlas.org/projects/detail?uuid=027c51c6-0719-469f-a7f5-640fe57cbece)
    2. [fc55bd4a-8694-4a28-9a35-59a685bda323](https://contribute.data.humancellatlas.org/projects/detail?uuid=fc55bd4a-8694-4a28-9a35-59a685bda323)

I can confirm that 2 of the 19 ignored projects are duplicates, so no need to import them into Ingest:
```
dcp4720991154984207     GSE75659	10.1038/ng.3678
dcp9566991661842434     GSE130973	10.1038/s42003-020-0922-4
```

* `dcp6037919259260882`: this project has the status `stalled` in the Dataset Tracking Sheet, [because it was deemed very low priority](https://github.com/ebi-ait/hca-ebi-wrangler-central/issues/103#issuecomment-664250397) It is a judgement call to include this project in the catalogue/Ingest or not.

* I am not sure if you want to include in the catalogue/Ingest the following 6 UCSC datasets:

```
dcp_id  project_short_name  data_accession  primary_wrangler
dcp1583639162870009     GSE84465    UCSC
dcp4005881213618518     GSE132802   UCSC
dcp8694575547617528     GSE130636   UCSC
dcp7134843558838905     GSE110499   UCSC
dcp0605441215849850     GSE36552    UCSC
dcp1901125925146315     GSE75659    UCSC

```

In summary, I can confirm that 12 of the 19 ignored projects can be ignored. The remaining 7 projects could be imported, but I think it would not be a great problem to ignore them for now.
UCSC wranglers usually upload spreadsheets with project metadata. Unless they know that their projects already exist in Ingest, and they actually know how to upload a spreadsheet so that it gets submitted to a specific existing project, then there is high probability for project duplication in Ingest if you automatically create the Ingest projects for the 6 UCSC projects from the dataset tracking sheet listed above.

@MightyAx : do my comments help? Shall I investigate **group 1** further? Please, let me know if you have any questions or concerns.

**Comment by clairerye** [comment]
This is great Ray. I think we should include 
dcp6037919259260882: this project has the status stalled in the Dataset Tracking Sheet, because it was deemed very low priority It is a judgement call to include this project in the catalogue/Ingest or not.
As although it is low priority it is eligible - I have added the DOI to the snapshot sheet.
I also think we should include the 6 UCSC datasets. But we should ensure we tell them the project has been reacted in ingest so we don't get duplications, as I agree, otherwise its almost guaranteed to hsppen.

**Comment by rays22** [comment]
@MightyAx : I will try to match up the technology and organ terms used in the dataset tracking spreadsheet with the correct ontology terms for Ingest import. I will use your tables to add the mappings. I will get back to you with the results and any problematic lines in the table that we might need to leave blank.
- [x] Resolve [Organ Ontologies that cannot be found](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=1923571015)
- [x] Resolve [Technology Ontologies that cannot be found](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=746177999)
- [x] Resolve [Ontologies that do not match exactly](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=781850240)

**Comment by tburdett** [comment]
I wanted to clarify that as part of this ticket it is **not** a requirement to resolve all ontology mapping problems in the first pass - this is less urgent and we can create another task to solve this in the second round. A recommendation of how we might resolve them would be useful (sounds like you're zooming in on manual review of the source data?).

There is **a strong requirement** to make sure we get all projects without ontology mapping problems into the production instance of the project catalogue, though, so we can review the interface, share with stakeholders, generate materials for the poster and tally all numbers to make sure we are only missing those we expect

**Comment by rays22** [comment]
@tburdett I am not sure if I understand your clarification.

I have already mapped the [single cell library construction](http://www.ebi.ac.uk/efo/EFO_0010183) technologies in this tab: <https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=746177999>
I have also mapped the problematic text terms from the dataset tracking sheet organ column to ontology terms here: <https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=1923571015>

Are you saying that I should stop mapping the dataset tracking sheet terms to ontology terms so that that these ontology terms could be imported with the corresponding projects into Ingest and the project catalogue?

**Comment by rays22** [comment]
I have updated the [Nearest Ontology Check](https://docs.google.com/spreadsheets/d/15imB5HErtY-wWaelnT_gxHRjWXyOQP6iagfPm0hU8sU/edit#gid=781850240) tab with the correct `text --> ontology term, ontology id` mappings.
